concent_api/core/models.py

```diff
@@ -73,3 +81,324 @@ class MessageAuth(Model):
+        # Check if all required related messages are not None in current state.
+        for stored_message_name, states in Subtask.REQUIRED_RELATED_MESSAGES_IN_STATES.items():
+            if self.state_enum in states and getattr(self, stored_message_name) is None:
+                raise ValidationError({
+                    stored_message_name: '{} cannot be None in state {}.'.format(
+                        stored_message_name,
+                        self.state,
+                    )
+                })
+
+    @property
+    def state_enum(self):
+        return Subtask.SubtaskState[self.state]
+
+    @property
+    def _current_state_enum(self):
+        return Subtask.SubtaskState[self._current_state_name]
+
+
+
+class PendingResponse(Model):
+    """
+    Stores information about messages to be returned from the `receive` or `receive-out-of-band` endpoint.
+    """
+
+    class ResponseType(ChoiceEnum):
+        ForceReportComputedTask         = 'ForceReportComputedTask'
+        ForceSubtaskResultsSettled      = 'ForceSubtaskResultsSettled'
+        AckReportComputedTask           = 'AckReportComputedTask'
+        RejectReportComputedTask        = 'RejectReportComputedTask'
+        VerdictReportComputedTask       = 'VerdictReportComputedTask'
+        ForceGetTaskResult              = 'ForceGetTaskResult'
+        ForceGetTaskResultRejected      = 'ForceGetTaskResultRejected'
+        ForceGetTaskResultFailed        = 'ForceGetTaskResultFailed'
+        ForceGetTaskResultUpload        = 'ForceGetTaskResultUpload'
+        ForceGetTaskResultDownload      = 'ForceGetTaskResultDownload'
+        ForceSubtaskResults             = 'ForceSubtaskResults'
+        SubtaskResultsSettled           = 'SubtaskResultsSettled'
+        ForceSubtaskResultsResponse     = 'ForceSubtaskResultsResponse'
+        SubtaskResultsRejected          = 'SubtaskResultsRejected'
+        ForcePaymentCommitted           = 'ForcePaymentCommitted'
+
+    class Queue(ChoiceEnum):
+        Receive             = 'receive'
+        ReceiveOutOfBand    = 'receive_out_of_band'
+
+    @property
+    def response_type_enum(self):
+        return PendingResponse.ResponseType[self.response_type]
+
+    response_type        = CharField(max_length = 32, choices = ResponseType.choices())
+    client               = ForeignKey(Client)
+    queue                = CharField(max_length = 32, choices = Queue.choices())
+
+    # TRUE if the client has already fetched the message.
+    delivered            = BooleanField(default = False)
+
+    subtask              = ForeignKey(Subtask, blank = True, null = True)
+    created_at           = DateTimeField(default = timezone.now)
+
+    def clean(self):
+        # payment_message can be included only if current state is ForcePaymentCommitted
+        if self.response_type != PendingResponse.ResponseType.ForcePaymentCommitted.name and self.payments.filter(pending_response__pk = self.pk).last() is not None:  # pylint: disable=no-member
```

Comment:
Is there any difference between `.last() is not None` and `.exists()`?

Reply:
Changed to `.exists()`